### PR TITLE
planner: add `model.ExtraPhysTblID` into indexScan.Schema()

### LINKED ISSUE #44984: execute select for update SQL with partition table panic
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t(a int, b int, index k(b)) PARTITION BY HASH(a) partitions 4;
insert into t(a, b) values (1,1),(2,2),(3,3),(4,4),(5,5),(6,6),(7,7),(8,8);
analyze table t;
select b from t use index(k) where b > 2 order by b limit 10 for update;
```

### 2. What did you expect to see? (Required)

not panic, return result

### 3. What did you see instead (Required)
```
mysql> select b from t use index(k) where b > 2 order by b limit 10 for update;
ERROR 1105 (HY000): runtime error: index out of range [2] with length 2
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```
mysql> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                             |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.3.0-alpha
Edition: Community
Git Commit Hash: 90c05ae4c648e55a150123e99b1f4164e3472047
Git Branch: heads/refs/tags/v7.3.0-alpha
UTC Build Time: 2023-06-26 14:26:39
GoVersion: go1.20.5
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #44984

Problem Summary: 
Sloved two problems, 
1. If we add `model.ExtraPhysTblID` in DataSource via `SelectLock`, we will not add it again in `AddExtraPhysTblIDColumn`, which will be missing this column in indexScan.Schema()
2. For `SelectLock`, `needPartitionHandle = flase` and `hasExtraCol = true` is exsits, we should allow it.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by mjonss — APPROVED** [review]

File: executor/distsql.go [hunk 1]
```diff
@@ -626,27 +626,28 @@ func (e *IndexLookUpExecutor) startWorkers(ctx context.Context, initBatchSize in
...
 	// TODO: fix global index related bugs later
-	if needPartitionHandle != ret && !e.index.Global {
-		return ret, errors.Errorf("Internal error, needPartitionHandle(%t) != ret(%t), tp(%d)", needPartitionHandle, ret, tp)
+	// For `SelectLock`, the `ExtraPhysTblID` will contained in schema,
+	// but it needn't partition handle to keep order.
+	if needPartitionHandle && !ret && !e.index.Global {
+		return ret, errors.Errorf("Internal error, needPartitionHandle != ret, tp(%d)", tp)
 	}
-	return ret, nil
+	return needPartitionHandle, nil
```
Comment (tiancaiamao) [inline executor/distsql.go:650]: `ret` is declared but not used?

Comment (Defined2014) [reply executor/distsql.go:650]: This is designed to check status, but this PR not perfect now. I still working on it.

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**